Lib/fractions.py

```diff
@@ -564,10 +564,25 @@ def __hash__(self):
         try:
             dinv = pow(self._denominator, -1, _PyHASH_MODULUS)
         except ValueError:
-            # ValueError means there is no modular inverse
+            # ValueError means there is no modular inverse.
             hash_ = _PyHASH_INF
         else:
-            hash_ = hash(abs(self._numerator)) * dinv % _PyHASH_MODULUS
+            # The general algorithm now specifies that the absolute value of
+            # the hash is
+            #    (|N| * dinv) % P
+            # where N is self._numerator and P is _PyHASH_MODULUS.  That's
+            # optimized here in two ways:  first, for a non-negative int i,
+            # hash(i) == i % P, but the int hash implementation doesn't need
+            # to divide, and is faster than doing % P explicitly.  So we do
+            #    hash(|N| * dinv)
+            # instead.  Second, N is unbounded, so its product with dinv may
+            # be arbitrarily expensive to compute.  The final answer is the
+            # same if we use the bounded |N| % P instead, which can again
+            # be done with an int hash() call.  If 0 <= i < P, hash(i) == i,
+            # so this nested hash() call wastes a bit of time making a
+            # redundant copy when |N| < P, but can save an arbitrarily large
+            # amount of computation for large |N|.
+            hash_ = hash(hash(abs(self._numerator)) * dinv)
         result = hash_ if self._numerator >= 0 else -hash_
         return -2 if result == -1 else result
```

Comment:
I think that you can replace this with `return result`. The slot wrapper for `tp_hash` already takes care to replace -1 by -2.

Reply:
Cool!  I didn't realize that, but it makes sense.  I'll make that change 😄

Reply:
Nope!  I had to revert that change - the test failed.  For `f = Fraction(-1)`, `hash(f)` did return -2, but `f.__hash__()` returned - 1 ☹️

Reply:
> `f.__hash__()` returned -1

Of course, `f.__hash__()` is just directly calling the Python function. But I personally wouldn't consider this a problem: you're not supposed to test `f.__hash__()`, what matters is `hash(f)`.

Reply:
It appears this was deliberately done, in

https://bugs.python.org/issue10356

In any case, I don't care to pursue it - the point of the recent `hash(Fraction)` fiddling was to improve speed, and the timing difference from removing the -1 special-casing was in the noise.

Reply:
Thanks for the pointer. This "hash equal to -1" thing really is an unfortunate side effect of the CPython implementation.